sql: separate out lease state from table version state

vivekmenezes opened the pull request:
The first commit is part of #16453 and should not be reviewed.

This PR gets us closer to allowing the LeaseManager to serve a table descriptor version for a prior version that does not need a lease.

vivekmenezes committed:
sql: replace LeaseState with tableVersionState

tableVersionState holds a table descriptor with its expiration time
and any information related to a lease it is holding on behalf of
the table version. For now a tableVersionState always holds a lease
but in the future only the latest two versions will hold a lease.

also changed leaseSet to tableSet representing a collection of
table versions.

vivekmenezes committed:
sql: separate leaseState from tableVersionState

This change moves the lease state out of the tableVersionState
so that we can store a tableVersionState that has a
version <= latest_version - 2 in the tableSet. Such a prior version
does not need a lease and will be represented by a tableVersionState
that does not have a leaseState

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16527)
<!-- Reviewable:end -->

vivekmenezes committed:
sql: unexport LeaseStore.Acquire/Release

andreimatei commented:
Review status: 0 of 23 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease.go, line 67 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv91k3L4zdUKiXesr:-Kmgv91k3L4zdUKiXess:byn0pse) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b00f000223906c71c0356528aa16a2d785d2281/pkg/sql/lease.go#L67)):*
> ```Go
> 	// Care must be taken to not modify it.
> 	sqlbase.TableDescriptor
> 	expiration parser.DTimestamp
> ```

how come this is removed from here? Descriptors without a lease also have an "expiration".

---

*[pkg/sql/lease.go, line 64 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgw_iNqOyA3iGOqVFq:-Kmgw_iNqOyA3iGOqVFr:bkycxj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L64)):*
> ```Go
> 	expiration parser.DTimestamp
> 
> 	testingKnobs LeaseStoreTestingKnobs
> ```

It bothered me that each `leaseState` holds a reference to the testing knobs (which one would think are meant for the `LeaseStore`); the only reason why they have this reference is for the `CanUseExpiredLeases` member, which is only set by one test (`TestTxnObeysLeaseExpiration`).
If you're so inclined, consider doing something else in that test and removing this knob. The way that test is testing what it wants to test seems quite funky to me, and I don't have any confidence that it's really testing anything, so I think it'd be a great thing if it were re-written regardless of this issue :). The test acquires a lease, then changes the expiration of that lease to be in the past, then tries to run a transaction using that expired lease and expects it to fail.

I think instead we could do one, or all, of the following things:
- test that getting a table descriptor (e.g. from a lease) affects a transaction's deadline
- test that a transaction's deadline is enforced when it tries to commit (this test doesn't have anything to do with SQL specifically). There's already a `TestEndTransactionDeadline` which sounds related.
- for an integration test, we should test a real scenario (rather than the dubious thing we're doing now) - push a transaction over the expiration of its leases, and check that a) a SNAPSHOT transaction (which would normally succeed), fails to commit and b) upon restart, both a SERIALIZABLE transaction and a SNAPSHOT one acquire a new lease.

And btw, since we're talking, should the "transaction deadline exceeded" error that that test expects be retryable? I think it should... Btw, for read-only transaction the deadline is enforced in `client.Txn`, and there the error is retryable.

---

*[pkg/sql/lease.go, line 69 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-KmgvXibt0x-yjl3YVHP:-KmgvXibt0x-yjl3YVHQ:bb39bvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L69)):*
> ```Go
> // tableVersionState holds the state for a table version. This includes
> // the lease information for a table version.
> // TODO(vivek): A node only needs to manage lease information on what it
> ```

is this TODO satisfied by this commit?

---

*[pkg/sql/lease.go, line 82 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv1Gf2f0x8LdNMXak:-Kmgv1GgLz_3mb4VselS:b-5a7haw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L82)):*
> ```Go
> 	// table name cache can have references to such tables since releasing a lease
> 	// and updating the cache is not atomic.
> 	released bool
> ```

should this `released` field move to the new `leaseState`?

---

*[pkg/sql/lease.go, line 84 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmh2z59Io6ZRh5Yc9qV:-Kmh2z59Io6ZRh5Yc9qW:b8wuzzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L84)):*
> ```Go
> 	released bool
> 
> 	// A lease for the table version if needed.
> ```

It'd be nice if the "if needed" part had more explanation. And perhaps spell out that the lease will be nil otherwise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Thanks for taking a look. I've postponed some of what you have suggested for the future. Thanks for taking a look!

---

Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/lease.go, line 67 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv91k3L4zdUKiXesr:-Kn-h5qlqJR8iOCqH4YS:bxcdtqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b00f000223906c71c0356528aa16a2d785d2281/pkg/sql/lease.go#L67)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come this is removed from here? Descriptors without a lease also have an "expiration".
</blockquote></details>

True, but in this PR we are still only supporting table descriptors with leases. In the future I suspect we would load up the table descriptor with ```version + 2``` and use its ModificationTime  as the expiration time so we might not need to store it here.

---

*[pkg/sql/lease.go, line 64 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgw_iNqOyA3iGOqVFq:-Kn-gbpUMdz0GV3mCdbk:b-moifzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L64)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It bothered me that each `leaseState` holds a reference to the testing knobs (which one would think are meant for the `LeaseStore`); the only reason why they have this reference is for the `CanUseExpiredLeases` member, which is only set by one test (`TestTxnObeysLeaseExpiration`).
If you're so inclined, consider doing something else in that test and removing this knob. The way that test is testing what it wants to test seems quite funky to me, and I don't have any confidence that it's really testing anything, so I think it'd be a great thing if it were re-written regardless of this issue :). The test acquires a lease, then changes the expiration of that lease to be in the past, then tries to run a transaction using that expired lease and expects it to fail.

I think instead we could do one, or all, of the following things:
- test that getting a table descriptor (e.g. from a lease) affects a transaction's deadline
- test that a transaction's deadline is enforced when it tries to commit (this test doesn't have anything to do with SQL specifically). There's already a `TestEndTransactionDeadline` which sounds related.
- for an integration test, we should test a real scenario (rather than the dubious thing we're doing now) - push a transaction over the expiration of its leases, and check that a) a SNAPSHOT transaction (which would normally succeed), fails to commit and b) upon restart, both a SERIALIZABLE transaction and a SNAPSHOT one acquire a new lease.

And btw, since we're talking, should the "transaction deadline exceeded" error that that test expects be retryable? I think it should... Btw, for read-only transaction the deadline is enforced in `client.Txn`, and there the error is retryable.

</blockquote></details>

I think TestTxnObeysLeaseExpiration is going to look quite different once I've implemented checking txn timestamp against the validity window, as opposed to depending on the txn from enforcing it. I'm inclined to leave this as is. I'm going to add a TODO to get rid of it once we've refashioned TestTxnObeysLeaseExpiration

---

*[pkg/sql/lease.go, line 69 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-KmgvXibt0x-yjl3YVHP:-Kn-hUfNcHvrhKp-tKf2:b-pxqto8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L69)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is this TODO satisfied by this commit?
</blockquote></details>

Not yet. I'd wait for the next PR that is going to add code that loads up tables for txns that cannot use the latest 2 versions.

---

*[pkg/sql/lease.go, line 82 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv1Gf2f0x8LdNMXak:-Kn-hjy89OSJxiDURut_:b-okm7ps) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L82)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

should this `released` field move to the new `leaseState`?
</blockquote></details>

No. If a lease is released the leaseState pointer within the tableVersionState is set to nil. This is needed for the case where we want to force the higher layer cache to
not use this table descriptor and refresh it to the latest lease. And we still do need it because we eventually are going to hold tableVersionState that do not have a lease
and thus have a nil lease.
 

---

*[pkg/sql/lease.go, line 84 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmh2z59Io6ZRh5Yc9qV:-Kn-nDf1ypd4ARyCDHg_:b-9cl5bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L84)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It'd be nice if the "if needed" part had more explanation. And perhaps spell out that the lease will be nil otherwise.
</blockquote></details>

Added a comment explaining when a lease is needed and when not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-:-Kn-nnvrdNLteBVc7rs9:bwnjy3n)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@andreimatei can I merge this. I'd like to keep moving on with the
implementation. Thanks!

On Thu, Jun 15, 2017, 2:51 PM Andrei Matei <notifications@github.com> wrote:

> Review status: 0 of 23 files reviewed at latest revision, 5 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/sql/lease.go, line 67 at r3
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv91k3L4zdUKiXesr:-Kmgv91k3L4zdUKiXess:byn0pse>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/8b00f000223906c71c0356528aa16a2d785d2281/pkg/sql/lease.go#L67>):*
>
> 	// Care must be taken to not modify it.
> 	sqlbase.TableDescriptor
> 	expiration parser.DTimestamp
>
> how come this is removed from here? Descriptors without a lease also have
> an "expiration".
> ------------------------------
>
> *pkg/sql/lease.go, line 64 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgw_iNqOyA3iGOqVFq:-Kmgw_iNqOyA3iGOqVFr:bkycxj3>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L64>):*
>
> 	expiration parser.DTimestamp
>
> 	testingKnobs LeaseStoreTestingKnobs
>
> It bothered me that each leaseState holds a reference to the testing
> knobs (which one would think are meant for the LeaseStore); the only
> reason why they have this reference is for the CanUseExpiredLeases
> member, which is only set by one test (TestTxnObeysLeaseExpiration).
> If you're so inclined, consider doing something else in that test and
> removing this knob. The way that test is testing what it wants to test
> seems quite funky to me, and I don't have any confidence that it's really
> testing anything, so I think it'd be a great thing if it were re-written
> regardless of this issue :). The test acquires a lease, then changes the
> expiration of that lease to be in the past, then tries to run a transaction
> using that expired lease and expects it to fail.
>
> I think instead we could do one, or all, of the following things:
>
>    - test that getting a table descriptor (e.g. from a lease) affects a
>    transaction's deadline
>    - test that a transaction's deadline is enforced when it tries to
>    commit (this test doesn't have anything to do with SQL specifically).
>    There's already a TestEndTransactionDeadline which sounds related.
>    - for an integration test, we should test a real scenario (rather than
>    the dubious thing we're doing now) - push a transaction over the expiration
>    of its leases, and check that a) a SNAPSHOT transaction (which would
>    normally succeed), fails to commit and b) upon restart, both a SERIALIZABLE
>    transaction and a SNAPSHOT one acquire a new lease.
>
> And btw, since we're talking, should the "transaction deadline exceeded"
> error that that test expects be retryable? I think it should... Btw, for
> read-only transaction the deadline is enforced in client.Txn, and there
> the error is retryable.
> ------------------------------
>
> *pkg/sql/lease.go, line 69 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-KmgvXibt0x-yjl3YVHP:-KmgvXibt0x-yjl3YVHQ:bb39bvc>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L69>):*
>
> // tableVersionState holds the state for a table version. This includes// the lease information for a table version.// TODO(vivek): A node only needs to manage lease information on what it
>
> is this TODO satisfied by this commit?
> ------------------------------
>
> *pkg/sql/lease.go, line 82 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv1Gf2f0x8LdNMXak:-Kmgv1GgLz_3mb4VselS:b-5a7haw>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L82>):*
>
> 	// table name cache can have references to such tables since releasing a lease
> 	// and updating the cache is not atomic.
> 	released bool
>
> should this released field move to the new leaseState?
> ------------------------------
>
> *pkg/sql/lease.go, line 84 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmh2z59Io6ZRh5Yc9qV:-Kmh2z59Io6ZRh5Yc9qW:b8wuzzr>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L84>):*
>
> 	released bool
>
> 	// A lease for the table version if needed.
>
> It'd be nice if the "if needed" part had more explanation. And perhaps
> spell out that the lease will be nil otherwise.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16527>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/16527#issuecomment-308834305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBJWiizxru6qCy_AZlV0ToP7wga3hks5sEX0_gaJpZM4N6Vg2>
> .
>

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/lease.go, line 82 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv1Gf2f0x8LdNMXak:-KnAJi4vabGUYkX4gyoJ:bmwdtn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L82)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

No. If a lease is released the leaseState pointer within the tableVersionState is set to nil. This is needed for the case where we want to force the higher layer cache to
not use this table descriptor and refresh it to the latest lease. And we still do need it because we eventually are going to hold tableVersionState that do not have a lease
and thus have a nil lease.
 
</blockquote></details>

I see...
I'd suggest renaming to `invalid` then (and, when setting it, consider zero-ing out the whole struct except this field).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-:-KnAK5JHRkknc1s4h2Ca:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: rename tableVersionState.released to invalid

vivekmenezes commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/lease.go, line 82 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-Kmgv1Gf2f0x8LdNMXak:-KnB9hKmLMrLg9C4LgJ2:b-1t8rpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1618903b4e7e8923666c4b035c925e229f6e4b2/pkg/sql/lease.go#L82)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I see...
I'd suggest renaming to `invalid` then (and, when setting it, consider zero-ing out the whole struct except this field).
</blockquote></details>

I've renamed it but unfortunately it runs some cleanup work after setting it to true so we can't blindly zero out the other fields

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16527#-:-KnBGR8sV-kPFhm3Imyh:bnfp4nl)*
<!-- Sent from Reviewable.io -->
